File: problem1.cpp
/*

// Time Complexity : O(M*N)
// Space Complexity : O(M*N) number of stack recursive calls
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this :


// Your code here along with comments explaining your approach
ITerate throught the matrix, find the 1 and then make all conntected 1 -> 0 . Increase the count by 1.
*/

#include<iostream>
#include<vector>
#include<queue>

using namespace std;

class Solution {
    void dfs(vector<vector<char>>& grid ,int x,int y,int m,int n){
        if(x>=0 && x<m && y>=0 && y<n && grid.at(x).at(y) == '1' ){
            grid.at(x).at(y) = '0';
            dfs(grid,x+1,y,m,n);
            dfs(grid,x-1,y,m,n);
            dfs(grid,x,y+1,m,n);
            dfs(grid,x,y-1,m,n);
        }
    }
public:
    int numIslands(vector<vector<char>>& grid) {
        int m = grid.size();
        int n = grid.at(0).size();
        int cnt{};
        for(int i{};i<m;++i){
            for(int j{};j<n;++j){
                if(grid.at(i).at(j) == '1'){
                    dfs(grid,i,j,m,n);
                    cnt++;
                }
                
            }
        }
        return cnt;
    }
};

File: Problem_1.js
// Problem1 (https://leetcode.com/problems/number-of-islands/)

// DFS
// TC: O(2mn)
// SC: O(mn) -> Recursion stack

// BFS
// TC: O(mn)
// SC: O(min(m,n))

// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No
// Your code here along with comments explaining your approach

let dirs;
let dfs = (grid, i, j) => {
    if (i < 0 || i >= grid.length || j < 0 || j > grid[0].length || grid[i][j] !== "1")
        return;
    // Set visited island to "2"
    grid[i][j] = "2";
    dirs.forEach(dir => {
        dfs(grid, i + dir[0], j + dir[1]);
    })
}

let bfs = (grid, i, j) => {
    // Queue will store the visited nodes
    let q = [[i, j]];
    grid[i][j] = "2";
    while (q.length > 0) {
        let node = q.shift();
        // Check all 4 directions
        dirs.forEach(dir => {
            let nr = node[0] + dir[0];
            let nc = node[1] + dir[1];
            // If a valid cell, change to 2
            if (nr >= 0 && nr < grid.length && nc >= 0 && nc < grid[0].length && grid[nr][nc] === "1") {
                grid[nr][nc] = "2";
                q.push([nr, nc]);
            }
        })
    }
}

/**
 * @param {character[][]} grid
 * @return {number}
 */
var numIslands = function (grid) {
    if (grid === null)
        return 0;

    let totalIslands = 0;
    dirs = [[1, 0], [0, 1], [-1, 0], [0, -1]];

    // Go through all the grid elements
    for (let i = 0; i < grid.length; i++) {
        for (let j = 0; j < grid[0].length; j++) {
            // If it is an island i.e "1", start searching for island
            if (grid[i][j] === "1") {
                bfs(grid, i, j);
                // dfs(grid, i, j);
                totalIslands++;
            }
        }
    }

    // reverting the grid
    for (let i = 0; i < grid.length; i++) {
        for (let j = 0; j < grid[0].length; j++) {
            if (grid[i][j] === "2") {
                grid[i][j] = "1"
            }
        }
    }

    return totalIslands;
};